pkg/retry/retry.go

```diff
@@ -104,61 +112,80 @@ func (s *Server) handleRetry() http.Handler {
 				return
 			}
 
-			logger.Infow("retrieve deliveries from GitHub", "checkpoint", cursor, "pageSize", len(deliveries))
-
-			// append to the total events counter
-			totalEventCount += len(deliveries)
+			logger.Infow("retrieve deliveries from GitHub", "cursor", cursor, "size", len(deliveries))
 
 			// update the cursor
 			cursor = res.Cursor
 
 			// for each failed delivery, redeliver
 			for i := 0; i < len(deliveries); i++ {
+				// append to the total events counter
+				totalEventCount += 1
+
 				event := deliveries[i]
+
+				// reached the last successful checkpoint, all events equal to and older
+				// than this one have already been processed
+				if lastCheckpoint == strconv.FormatInt(*event.ID, 10) {
+					processed = true
+					break
+				}
+
 				// check payload and see if its been successfully delivered, if so skip over it
 				if *event.StatusCode >= 200 && *event.StatusCode <= 299 {
 					continue
 				}
 
+				// append to the total failure events counter
+				failedEventCount += 1
+
 				logger.Infow("redeliver failed event", "eventID", *event.ID, "guid", *event.GUID)
 
 				if err := s.github.RedeliverEvent(ctx, *event.ID); err != nil {
-					logger.Errorw("failed to call RedeliverEvent",
+					logger.Errorw("failed to call RedeliverEvent, stop processing",
```

Comment:
Should we consider rate limiting here? Just curious about downtime and we have to send a ton of these, we could hit rate limits and need to throttle ourselves. Rather than failing completely we could back off (or check headers for the next time we can send).

This could be a follow up.

Reply:
Good call, it looks like we can anticipate a [422 from the redeliver API](https://docs.github.com/en/rest/orgs/webhooks?apiVersion=2022-11-28#redeliver-a-delivery-for-an-organization-webhook). Created an issue for this https://github.com/abcxyz/github-metrics-aggregator/issues/64.